### PR TITLE
Disallow duplicate `dynamic_templates` names

### LINKED ISSUE #28988: Dynamic templates should be deduplicated
**Bug Report / Feature request**:

Check for identical entries in the list of dynamic templates and deduplicate or reject them to avoid bloating.

**Describe the issue**:

It is currently possible to have the same dynamic template definition appended multiple times in the dynamic templates list of a mapping or a template, which can result to significant performance degradation.

See example [template](https://github.com/elastic/elasticsearch/files/1802117/Metricbeat_template.gz) of an actual problematic 300.000 lines+ configuration installed in a cluster due to a repetition when putting the dynamic templates. This eventually resulted in a slowdown, tasks piling up, causing rejected bulks etc.

**Elasticsearch version**: issue exists up to current - 6.2.2

**Steps to reproduce**:

PUT the same dynamic template definition multiple times when defining a mapping or a template:

```
PUT my_index
{
  "mappings": {
    "_doc": {
      "dynamic_templates": [
        {
          "docker.container.labels": {
            "mapping": {
              "type": "keyword"
            },
            "match_mapping_type": "string",
            "path_match": "docker.container.labels.*"
          }
        },
        {
          "docker.container.labels": {
            "mapping": {
              "type": "keyword"
            },
            "match_mapping_type": "string",
            "path_match": "docker.container.labels.*"
          }
        }
      ]
    }
  }
}
```

### PR DESCRIPTION
**Pull request by cbuescher**
Currently the same `dynamic_templates` name can be used more than once in the
same mappings file. This most certainly is unintentional and can create issues like
those reported in #28988. Furthermore, when templates are matched later in
`RootObjectMapper#findTemplate`, it looks like we simply pick the first matching
one we see in the input array, so using the same template name twice sounds like
an error prone idea anyway. This change checks for duplicate names on parsing
and throws an error if we encounter the same name twice.

Closes #28988

### CODE REVIEW

**Comment by cbuescher** [comment]
@jtibshirani I found this by accident since I was reviewing a PR in the same area earlier and thought this might be a simple fix. Let me know if this makes sense to you or if my thinking around the issue was a bit too shortsighted.

File: server/src/test/java/org/elasticsearch/index/mapper/DynamicMappingTests.java [hunk 1]
```diff
@@ -607,7 +607,7 @@ public void testDateDetectionInheritsFormat() throws Exception {
                         .endObject()
                     .endObject()
                     .startObject()
-                        .startObject("dates")
+                        .startObject("dates_with_explicit_format")
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/mapper/DynamicMappingTests.java]: After having to change this test I'm not entirely sure anymore about the approach, in particular whether this test was lazily "missusing" a leniency in our code by using the name twice or wether this should work. In the end both templates seem to match different fields, so I'm not sure why the test used the same name?

**Review by jtibshirani — COMMENTED** [review]
@cbuescher since commenting on #28988, I now think that some users are relying on this behavior (where dynamic templates are not disallowed/ deduplicated). For example, we recently found out in https://github.com/elastic/elasticsearch/issues/29200#issuecomment-534028143 that APM creates multiple dynamic with the same name, but with different `match_mapping_type` values.

Would it make sense to discuss as a team to make sure we're okay with the direction? A note that if we do go this direction, I think we'd need a deprecation period before we start throwing an error, as there are users taking advantage of the leniency.

**Comment by cbuescher** [comment]
> Would it make sense to discuss as a team to make sure we're okay with the direction

Absolutely, I wasn't aware of the other issue (#29200). From a first glance, it seems using the same dynamic template name there is problematic for similar reasons, but it makes sense to look at this from another angle again. We can close this PR for now and I will mark the original issue for discussion.

**Pull request closed without merging**